pkg/statedb2/txn.go

```diff
@@ -0,0 +1,369 @@
+
+import (
+	"bufio"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	iradix "github.com/hashicorp/go-immutable-radix/v2"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+type txn struct {
+	db               *DB
+	rootReadTxn      *iradix.Txn[tableEntry]            // read transaction onto the tree of tables
+	lastIndexReadTxn lastIndexReadTxn                   // memoized result of the last indexReadTxn()
+	writeTxns        map[tableIndex]*iradix.Txn[object] // opened per-index write transactions
+	tables           map[TableName]*tableEntry          // table entries being modified
+	smus             lock.SortableMutexes               // the (sorted) table locks
+}
+
+type tableIndex struct {
+	table TableName
+	index IndexName
+}
+
+type lastIndexReadTxn struct {
+	table TableName
+	index IndexName
+	txn   *iradix.Txn[object]
+}
+
+var zeroTxn = txn{}
+
+func revisionKey(rev uint64, idKey []byte) []byte {
+	buf := make([]byte, 8+len(idKey))
+	binary.BigEndian.PutUint64(buf, rev)
+	copy(buf[8:], idKey)
+	return buf
+}
+
+// txn fulfills the ReadTxn/WriteTxn interface.
+func (txn *txn) getTxn() *txn {
+	return txn
+}
+
+func (txn *txn) GetRevision(name TableName) Revision {
+	if table, ok := txn.tables[name]; ok {
+		// This is a write transaction preparing to modify the table with a
+		// new revision.
+		return table.revision
+	}
+
+	// This is either a read transaction, or a write transaction to tables
+	// other than this table. Look up the revision from the index.
+	table, ok := txn.rootReadTxn.Get([]byte(name))
+	if !ok {
+		panic("BUG: Table " + name + " not found")
+	}
+	return table.revision
+}
+
+// indexReadTxn returns a transaction to read from the specific index.
+func (txn *txn) indexReadTxn(name TableName, index IndexName) *iradix.Txn[object] {
+	if txn.writeTxns != nil {
+		indexTxn, ok := txn.writeTxns[tableIndex{name, index}]
+		if ok {
+			return indexTxn.Clone()
+		}
+		if _, ok := txn.tables[name]; ok {
+			// We're writing into this table, create a write transaction
+			// instead.
+			return txn.indexWriteTxn(name, index).Clone()
+		}
+	}
+
+	if txn.lastIndexReadTxn.table == name && txn.lastIndexReadTxn.index == index {
+		return txn.lastIndexReadTxn.txn
+	}
+
+	table, ok := txn.rootReadTxn.Get([]byte(name))
+	if !ok {
+		panic("BUG: Table '" + name + "' not found")
+	}
+	indexTree, ok := table.indexes.Get([]byte(index))
+	if !ok {
+		panic(fmt.Sprintf("BUG: Index %s/%s not found", name, index))
+	}
+
+	indexTxn := indexTree.Txn()
+	txn.lastIndexReadTxn = lastIndexReadTxn{table: name, index: index, txn: indexTxn}
+	return indexTxn
+}
+
+// indexWriteTxn returns a transaction to read/write to a specific index.
+// The created transaction is memoized and used for subsequent reads and/or writes.
+func (txn *txn) indexWriteTxn(name TableName, index IndexName) *iradix.Txn[object] {
+	if indexTreeTxn, ok := txn.writeTxns[tableIndex{name, index}]; ok {
+		return indexTreeTxn
+	}
+	table, ok := txn.tables[name]
```

Comment:
there's a couple of these panics based on not finding the table; I think there's `getTable` for this already